src/js/common/util.js

```diff
@@ -0,0 +1,161 @@
+};
+
+
+/**
+ * npsetApplications - Goes through all the contexts in a flat prefs set
+ * and pulls out the flat pref URI and app ID. Returns all the apps a
+ * user has settings.
+ *
+ * @param prefs (Object) Transformed flat preferences set
+ * @return (Array) An array of objects structured as the following: (RWG example)
+ *     {
+ *         uri: "http://registry.gpii.net/applications/com.texthelp.readWriteGold",
+ *         appId: "com.texthelp.readWriteGold"
+ *     }
+ */
+gpii.devpmt.npsetApplications = function (prefs) {
+    var apps = {};
+    fluid.each(prefs.contexts, function (context) {
+        fluid.each(context.preferences, function (prefBody, prefKey) {
+            if (prefKey.startsWith("http://registry.gpii.net/applications")) {
+                if (!apps[prefKey]) {
+                    apps[prefKey] = {
+                        uri: prefKey,
+                        appId: prefKey.split(/\//).pop(),
+                        settingKeys: []
+                    };
+                }
+                // Add any keys that may or may not be in one of the contexts
+                fluid.each(prefBody, function (settingBody, settingKey) {
+                    if (apps[prefKey].settingKeys.indexOf(settingKey) < 0) {
+                        apps[prefKey].settingKeys.push(settingKey);
+                    }
+                });
+            }
+        });
+    });
+    return fluid.values(apps);
+};
+
+/**
+ * Query a lunr index in our case insensitive, match anything style that
+ * is used for the various list/API filter boxes that are typically rendered
+ * at the top of a list or table. This routine is meant to encapsulate
+ * any of the oddities of lunr search strings, and necessary usage of
+ * various wildcards for list filtering use cases.
+ *
+ * @param {Object} lunrIndex The existing lunr index to query.
+ * @param {String} filterString The string used to filter the results.
+ * @return {Array} The matching lunr results.
+ */
+gpii.devpmt.lunrListFilterSearch = function (lunrIndex, filterString) {
+    // TODO escape entire string so search keywords like 'not' can be used
+    return lunrIndex.query(function (q) {
+        // The lunr querystring documentation isn't great, but apparently
+        // lowercasing the terms does a case insensitive search.
+        var lowerFilterString = filterString.toLowerCase();
+        q.term(lowerFilterString);
+        q.term("*" + lowerFilterString);
+        q.term("*" + lowerFilterString + "*");
+    });
+};
+
+/**
+ * Goes through a set of solutions, and for a productId and term,
+ * finds the supportedSetting entry (there can be multiple blocks of them)
+ * and returns it.
+ */
+gpii.devpmt.findProductSettingMetadata = function ( solutions, productId, term ) {
+    var solution = solutions[productId];
+    var togo = {};
+    fluid.each(solution.settingsHandlers, function (settingHandler) {
+        fluid.each(settingHandler.supportedSettings, function (supportedSetting, key) {
```

Comment:
There will indeed be multiple such terms, why just return the last one? Note that with recent work from Kaspar and Astea, the structure of these prefs documents will be quite different.

Reply:
I don't *think* any instance of a term occurs more than once, the issue here is that all the settings and their schemas are split across potentially multiple sets of ini files, registry handlers, etc, so you have to go through all of them to find what you're looking for.  That's what I meant by multiple. Multiple sets of settings handlers to go through.